src/v/iceberg/snapshot_json.cc

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/snapshot_json.h"
+
+#include "container/chunked_hash_map.h"
+#include "iceberg/json_utils.h"
+#include "iceberg/snapshot.h"
+#include "json/document.h"
+#include "model/timestamp.h"
+#include "strings/string_switch.h"
+
+namespace iceberg {
+
+namespace {
+
+constexpr std::string_view operation_to_str(snapshot_operation o) {
+    switch (o) {
+    case snapshot_operation::append:
+        return "append";
+    case snapshot_operation::replace:
+        return "replace";
+    case snapshot_operation::overwrite:
+        return "overwrite";
+    case snapshot_operation::delete_data:
+        return "delete";
+    }
+}
+
+snapshot_operation operation_from_str(const ss::sstring& operation_str) {
+    return string_switch<snapshot_operation>(operation_str)
+      .match(
+        operation_to_str(snapshot_operation::append),
+        snapshot_operation::append)
+      .match(
+        operation_to_str(snapshot_operation::replace),
+        snapshot_operation::replace)
+      .match(
+        operation_to_str(snapshot_operation::overwrite),
+        snapshot_operation::overwrite)
+      .match(
+        operation_to_str(snapshot_operation::delete_data),
+        snapshot_operation::delete_data);
+}
+
+} // namespace
+
+snapshot parse_snapshot(const json::Value& v) {
+    auto id = parse_required_i64(v, "snapshot-id");
+    std::optional<snapshot_id> parent_id;
+    auto parent_id_opt = parse_optional_i64(v, "parent-snapshot-id");
+    if (parent_id_opt.has_value()) {
+        parent_id.emplace(parent_id_opt.value());
+    }
+    auto timestamp_ms = parse_required_i64(v, "timestamp-ms");
+    auto manifest_list_path = parse_required_str(v, "manifest-list");
+    auto schema_id_opt = parse_optional_i32(v, "schema-id");
+    std::optional<schema::id_t> schema_id;
+    if (schema_id_opt.has_value()) {
+        schema_id.emplace(schema_id_opt.value());
+    }
+    const auto& summary_json = parse_required(v, "summary");
+    if (!summary_json.IsObject()) {
+        throw std::invalid_argument(fmt::format(
+          "Expected JSON object to parse field 'summary': {}",
+          summary_json.GetType()));
+    }
+    std::optional<snapshot_operation> operation;
+    chunked_hash_map<ss::sstring, ss::sstring> other_map;
+    for (const auto& m : summary_json.GetObject()) {
+        if (!m.name.IsString() || !m.value.IsString()) {
```

Comment:
Would a schema validator be useful for describing the requirements on this object (the entire snapshot and all inner fields etc)? 

Might be too much work for the small number of fields as it is right now, but can be useful to do single pass validation and documenting the expected structure in future.

Reply:
Yea it's a good thing to keep in mind, but I'll hold off for now. One thing I'm unsure of is if/how this JSON will evolve. IIUC the catalog is relatively new, and having a bit more flexibility to deal with JSON schema changes seems like a plus for now.

Reply:
>ne thing I'm unsure of is if/how this JSON will evolve.

using a schema validator may be worth it even if the schema is evolving since it allows you to remove all of the inline error checking and other semantic validation.

Reply:
Yea i think I'll revisit this once the core pieces of the Iceberg metadata path are in tact.